cloudconfig/podcfg/podcfg.go

```diff
@@ -0,0 +1,464 @@
+	"github.com/juju/juju/juju/paths"
+	"github.com/juju/juju/state/multiwatcher"
+	coretools "github.com/juju/juju/tools"
+)
+
+var logger = loggo.GetLogger("juju.cloudconfig.podcfg")
+
+// ControllerPodConfig represents initialization information for a new juju caas controller pod.
+type ControllerPodConfig struct {
+	// Tags is a set of tags/labels to set on the Pod, if supported. This
+	// should be populated using the PodLabels method in this package.
+	Tags map[string]string
+
+	// Bootstrap contains bootstrap-specific configuration. If this is set,
+	// Controller must also be set.
+	Bootstrap *BootstrapConfig
+
+	// Controller contains controller-specific configuration. If this is
+	// set, then the instance will be configured as a controller pod.
+	Controller *ControllerConfig
+
+	// APIInfo holds the means for the new pod to communicate with the
+	// juju state API. Unless the new pod is running a controller (Controller is
+	// set), there must be at least one controller address supplied.
+	// The entity name must match that of the pod being started,
+	// or be empty when starting a controller.
+	APIInfo *api.Info
+
+	// ControllerTag identifies the controller.
+	ControllerTag names.ControllerTag
+
+	// MachineNonce is set at provisioning/bootstrap time and used to
+	// ensure the agent is running on the correct pod.
+	MachineNonce string
+
+	// tools is the list of juju tools used to install the Juju agent
+	// on the new pod. Each of the entries in the list must have
+	// identical versions and hashes, but may have different URLs.
+	tools coretools.List
+
+	// DataDir holds the directory that juju state will be put in the new
+	// instance.
+	DataDir string
+
+	// LogDir holds the directory that juju logs will be written to.
+	LogDir string
+
+	// MetricsSpoolDir represents the spool directory path, where all
+	// metrics are stored.
+	MetricsSpoolDir string
+
+	// Jobs holds what machine jobs to run.
+	Jobs []multiwatcher.MachineJob
+
+	// MachineId identifies the new machine.
+	MachineId string
+
+	// AgentEnvironment defines additional configuration variables to set in
+	// the pod agent config.
+	AgentEnvironment map[string]string
+
+	// Series represents the pod series.
+	Series string
+
+	// MachineAgentServiceName is the init service name for the Juju agent(k8s service name but not systemd service).
+	MachineAgentServiceName string
+}
+
+// BootstrapConfig represents bootstrap-specific initialization information
+// for a new juju caas pod. This is only relevant for the bootstrap pod.
+type BootstrapConfig struct {
+	instancecfg.BootstrapConfig
+}
+
+// ControllerConfig represents controller-specific initialization information
+// for a new juju caas pod. This is only relevant for controller pod.
+type ControllerConfig struct {
+	instancecfg.ControllerConfig
+}
+
+// AgentConfig returns an agent config.
+func (cfg *ControllerPodConfig) AgentConfig(
+	tag names.Tag,
+	toolsVersion version.Number,
```

Comment:
Rename to jujuVersion.
But anyway it isn't needed  - the value in agent conf can be empty if that's allowed.
We only use this to perform in place upgrades of jujud on a machine and for k8s, we'll replace the whole pod container